### PR TITLE
Tighten comparison of skolem types

### PR DESCRIPTION
**Pull request by odersky**
Skolem types are =:= only if they are reference-equal. Two skolem types
with the same underlying type are not necessarily equal.

Tests continue to run under this tightened definition.

No review necessary yet. More commits will follow.

### CODE REVIEW

File: src/dotty/tools/dotc/core/TypeOps.scala [hunk 1]
```diff
@@ -12,16 +12,25 @@ import ast.tpd._
 
 trait TypeOps { this: Context => // TODO: Make standalone object.
 
+  final def asSeenFrom(tp: Type, pre: Type, cls: Symbol): Type = {
+    val m = if (pre.isStable || !ctx.phase.isTyper) null else new AsSeenFromMap(pre, cls)
+    var res = asSeenFrom(tp, pre, cls, m)
+    if (m != null && m.unstable) asSeenFrom(tp, SkolemType(pre), cls) else res
+  }
+
   final def asSeenFrom(tp: Type, pre: Type, cls: Symbol, theMap: AsSeenFromMap): Type = {
```
Comment (smarter) [inline src/dotty/tools/dotc/core/TypeOps.scala]: This method and its parameters should be documented.

File: src/dotty/tools/dotc/core/TypeOps.scala [hunk 2]
```diff
@@ -12,16 +12,25 @@ import ast.tpd._
...
 
     def toPrefix(pre: Type, cls: Symbol, thiscls: ClassSymbol): Type = /*>|>*/ ctx.conditionalTraceIndented(TypeOps.track, s"toPrefix($pre, $cls, $thiscls)") /*<|<*/ {
       if ((pre eq NoType) || (pre eq NoPrefix) || (cls is PackageClass))
         tp
-      else if (thiscls.derivesFrom(cls) && pre.baseTypeRef(thiscls).exists)
+      else if (thiscls.derivesFrom(cls) && pre.baseTypeRef(thiscls).exists) {
+        if (theMap != null && theMap.currentVariance <= 0 && !pre.isStable)
```
Comment (smarter) [inline src/dotty/tools/dotc/core/TypeOps.scala]: It'd be helpful to add a comment explaining what this condition is for exactly, and why it causes us to mutate something.

File: src/dotty/tools/dotc/core/TypeOps.scala [hunk 3]
```diff
@@ -33,7 +42,19 @@ trait TypeOps { this: Context => // TODO: Make standalone object.
         case tp: NamedType =>
           val sym = tp.symbol
           if (sym.isStatic) tp
-          else tp.derivedSelect(asSeenFrom(tp.prefix, pre, cls, theMap))
+          else {
+            val prevStable = theMap == null || !theMap.unstable
+            val pre1 = asSeenFrom(tp.prefix, pre, cls, theMap)
+            if (theMap != null && theMap.unstable && prevStable) {
```
Comment (smarter) [inline src/dotty/tools/dotc/core/TypeOps.scala]: A comment for this condition would be nice too

File: src/dotty/tools/dotc/core/TypeOps.scala [hunk 4]
```diff
@@ -12,16 +12,25 @@ import ast.tpd._
...
 
+  final def asSeenFrom(tp: Type, pre: Type, cls: Symbol): Type = {
+    val m = if (pre.isStable || !ctx.phase.isTyper) null else new AsSeenFromMap(pre, cls)
+    var res = asSeenFrom(tp, pre, cls, m)
+    if (m != null && m.unstable) asSeenFrom(tp, SkolemType(pre), cls) else res
+  }
+
   final def asSeenFrom(tp: Type, pre: Type, cls: Symbol, theMap: AsSeenFromMap): Type = {
 
     def toPrefix(pre: Type, cls: Symbol, thiscls: ClassSymbol): Type = /*>|>*/ ctx.conditionalTraceIndented(TypeOps.track, s"toPrefix($pre, $cls, $thiscls)") /*<|<*/ {
```
Comment (smarter) [inline src/dotty/tools/dotc/core/TypeOps.scala]: This inner method and its parameters should be documented too.

File: src/dotty/tools/dotc/core/TypeOps.scala [hunk 5]
```diff
@@ -54,6 +75,105 @@ trait TypeOps { this: Context => // TODO: Make standalone object.
 
   class AsSeenFromMap(pre: Type, cls: Symbol) extends TypeMap {
     def apply(tp: Type) = asSeenFrom(tp, pre, cls, this)
+    def currentVariance = variance
+    var unstable = false
```
Comment (smarter) [inline src/dotty/tools/dotc/core/TypeOps.scala]: The field `unstable` should have a comment explaining what it's for exactly.

**Comment by odersky** [comment]
The PR is now complete. I am not sure about the documentation parts. This whole th

**Comment by odersky** [comment]
Rebased to master

File: src/dotty/tools/dotc/core/tasty/TastyFormat.scala [hunk 6]
```diff
@@ -103,10 +103,10 @@ Standard-Section: "ASTs" TopLevelStat*
                   TERMREFpkg            fullyQualified_NameRef
                   TERMREF               possiblySigned_NameRef qual_Type
                   THIS                  clsRef_Type
-                  SKOLEMtype            refinedType_ASTRef
+                  REFINEDthis           refinedType_ASTRef
+                  SKOLEMtype            Type_ASTRef
```
Comment (smarter) [inline src/dotty/tools/dotc/core/tasty/TastyFormat.scala]: When can Skolems appear in typed trees?

**Comment by smarter** [comment]
Could this new `SkolemType` be used to replace `ctx.newSkolem`(https://github.com/lampepfl/dotty/blob/master/src/dotty/tools/dotc/core/Symbols.scala#L294-L297) or are they still different?

**Comment by odersky** [comment]
@smarter They are still different. In Applications, we need a TermRef for the method reference, which is what the newSkolem enables. SkolemType does not provide us with a method symbol.

File: src/dotty/tools/dotc/core/tasty/TastyFormat.scala (hunk 6 above)
Comment (odersky) [inline src/dotty/tools/dotc/core/tasty/TastyFormat.scala]: They will be removed in a separate PR (#638).

**Comment by odersky** [comment]
Would be great to see a review of this so we can get this in.

**Comment by odersky** [comment]
I am merging now, because the commit fixes an important unsoundness problem in Dotty. The longer we leave this open the more likely it is that other commits will introduce unsound code.
